// Copyright (c) 2023 dingodb.com, Inc. All Rights Reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

import "common.proto";

package dingodb.pb.error;

option java_package = "io.dingodb.error";
option cc_generic_services = true;

enum Errno {
  OK = 0;

  // universal [10000, 20000)
  EINTERNAL = 10000;
  ECONTEXT = 10001;
  EILLEGAL_PARAMTETERS = 10010;
  EKEY_NOTFOUND = 10011;
  EKEY_FORMAT = 10012;
  EKEY_EMPTY = 10013;
  EKEY_EXIST = 10014;
  ENOT_SUPPORT = 10100;
  EKEY_OUT_OF_RANGE = 10101;
  EKEY_SPLIT = 10102;

  ESERVICE_TYPE_NOT_FOUND = 10200;
  ESERVICE_METHOD_NOT_FOUND = 10201;

  // store [20000, 30000)
  ESTORE_NOTEXIST_RAFTENGINE = 20000;
  ESTORE_INVALID_CF = 20001;
  ESCAN_NOTFOUND = 20002;

  // meta [30000, 40000)
  ESCHEMA_EXISTS = 30000;
  ESCHEMA_NOT_FOUND = 30001;
  ESCHEMA_NOT_EMPTY = 30002;
  ETABLE_EXISTS = 30003;
  ETABLE_NOT_FOUND = 30004;
  ETABLE_DEFINITION_ILLEGAL = 30005;
  ETABLE_REGION_CREATE_FAILED = 30006;
  ETABLE_REGION_DROP_FAILED = 30007;
  ETABLE_METRICS_FAILED = 30008;
  // auto increment
  EAUTO_INCREMENT_WHILE_CREAT_TABLE = 30101;
  EAUTO_INCREMENT_EXIST = 30102;

  // coordinator [40000, 50000)
  ESPLIT_STATUS_ILLEGAL = 40000;
  ESPLIT_PEER_NOT_MATCH = 40001;
  EMERGE_STATUS_ILLEGAL = 40002;
  EMERGE_PEER_NOT_MATCH = 40003;
  EMERGE_RANGE_NOT_MATCH = 40004;
  ECHANGE_PEER_STATUS_ILLEGAL = 40005;
  ECHANGE_PEER_SIZE_ILLEGAL = 40006;
  ESTORE_NOT_FOUND = 40007;
  EREGION_CMD_ONGOING_CONFLICT = 40008;
  ESTORE_IN_USE = 40009;
  EKEYRING_ILLEGAL = 40010;
  EEXECUTOR_NOT_FOUND = 40011;
  EUSER_ALREADY_EXIST = 40012;
  EUSER_NOT_EXIST = 40013;
  EOP_NOT_SUPPORTED = 40014;
  ECHANGE_PEER_UNABLE_TO_REMOVE_LEADER = 40015;
  ETASK_LIST_CONFLICT = 40016;

  // raft [50000, 60000)
  ERAFT_INIT = 50000;
  ERAFT_NOTNODE = 50001;
  ERAFT_NOTLEADER = 50002;
  ERAFT_COMMITLOG = 50003;
  ERAFT_SAVE_SNAPSHOT = 50010;
  ERAFT_LOAD_SNAPSHOT = 50011;

  // region [60000, 70000)
  EREGION_ALREADY_EXIST = 60000;
  EREGION_NOT_FOUND = 60001;
  EREGION_UNAVAILABLE = 60002;
  EREGION_STATE = 60003;
  EREGION_REPEAT_COMMAND = 60004;
  EREGION_ALREADY_DELETED = 60005;
  EREGION_ALREADY_SPLIT = 60006;
  EREGION_ALREADY_MERGED = 60007;
  EREGION_ALREADY_PEER_CHANGED = 60008;
  EREGION_REDIRECT = 60009;
}

message Error {
  Errno errcode = 1;
  string errmsg = 2;

  dingodb.pb.common.Location leader_location = 3;
}
